File: src/main/java/com/meepoffaith/hextra/init/Patterns.java
package com.meepoffaith.hextra.init;

import at.petrak.hexcasting.api.casting.ActionRegistryEntry;
import at.petrak.hexcasting.api.casting.castables.Action;
import at.petrak.hexcasting.api.casting.castables.OperationAction;
import at.petrak.hexcasting.api.casting.math.HexDir;
import at.petrak.hexcasting.api.casting.math.HexPattern;
import at.petrak.hexcasting.common.lib.hex.HexActions;
import com.meepoffaith.hextra.HextraPatterns;
import com.meepoffaith.hextra.casting.actions.MathActions.DegRad;
import com.meepoffaith.hextra.casting.actions.MathActions.RadDeg;
import com.meepoffaith.hextra.casting.actions.VecActions.VecNegOne;
import com.meepoffaith.hextra.casting.actions.VecActions.VecOne;
import net.minecraft.registry.Registry;
import net.minecraft.util.Identifier;

import java.util.LinkedHashMap;
import java.util.Map;

import static com.meepoffaith.hextra.init.Arithmetics.*;

public class Patterns{
    private static final Map<Identifier, ActionRegistryEntry> PATTERNS = new LinkedHashMap<>();

    public static final HexPattern DEG_TO_RAD = make("deg_to_rad", "qqqqqdwdq", HexDir.WEST, new DegRad());
    public static final HexPattern RAD_TO_DEG = make("rad_to_deg", "qdwdqqqqq", HexDir.NORTH_EAST, new RadDeg());

    //Come on, Elise!
    public static final HexPattern VEC_ONE = make("haha_ha_one", "qqqqqeq", HexDir.NORTH_WEST, new VecOne());
    public static final HexPattern VEC_NEGONE = make("eno_ah_ahah", "eeeeeqq", HexDir.SOUTH_WEST, new VecNegOne());

    public static HexPattern ROT_ABOUT_X = make("rot_about_x", "aaqqqqqea", HexDir.SOUTH_WEST);
    public static HexPattern ROT_ABOUT_Y = make("rot_about_y", "aaqqqqqew", HexDir.SOUTH_WEST);
    public static HexPattern ROT_ABOUT_Z = make("rot_about_z", "aaqqqqqed", HexDir.SOUTH_WEST);
    public static HexPattern CONSTRUCT_ABOUT_X = make("cons_about_x", "daqqqqqea", HexDir.NORTH_WEST);
    public static HexPattern CONSTRUCT_ABOUT_Y = make("cons_about_y", "daqqqqqew", HexDir.NORTH_WEST);
    public static HexPattern CONSTRUCT_ABOUT_Z = make("cons_about_z", "daqqqqqed", HexDir.NORTH_WEST);
    public static HexPattern NORMALIZE = make("normalize", "eeeeedww", HexDir.SOUTH_WEST);

    public static void init(){
        for(Map.Entry<Identifier, ActionRegistryEntry> entry : PATTERNS.entrySet()){
            Registry.register(HexActions.REGISTRY, entry.getKey(), entry.getValue());
        }
    }

    private static HexPattern make(String name, String sig, HexDir dir, Action act){
        HexPattern pattern = HexPattern.fromAngles(sig, dir);
        PATTERNS.put(new Identifier(HextraPatterns.MOD_ID, name), new ActionRegistryEntry(pattern, act));
        return pattern;
    }

    private static HexPattern make(String name, String sig, HexDir dir){
        HexPattern pattern = HexPattern.fromAngles(sig, dir);
        PATTERNS.put(new Identifier(HextraPatterns.MOD_ID, name), new ActionRegistryEntry(pattern, new OperationAction(pattern)));
        return pattern;
    }
}

File: src/main/kotlin/com/meepoffaith/hextrapats/init/Patterns.kt
package com.meepoffaith.hextrapats.init

import at.petrak.hexcasting.api.casting.ActionRegistryEntry
import at.petrak.hexcasting.api.casting.castables.Action
import at.petrak.hexcasting.api.casting.castables.OperationAction
import at.petrak.hexcasting.api.casting.castables.SpecialHandler
import at.petrak.hexcasting.api.casting.math.HexDir
import at.petrak.hexcasting.api.casting.math.HexPattern
import at.petrak.hexcasting.common.lib.hex.HexActions
import at.petrak.hexcasting.xplat.IXplatAbstractions
import com.meepoffaith.hextrapats.HextraPats
import com.meepoffaith.hextrapats.casting.actions.NoConsOperationAction
import com.meepoffaith.hextrapats.casting.actions.eval.OpConditionalEval
import com.meepoffaith.hextrapats.casting.actions.eval.OpConditionalHalt
import net.minecraft.registry.Registry

object Patterns {
    /*
    val ROT_ABOUT_X = register("rot_about_x", "aaqqqqqea", HexDir.SOUTH_WEST)
    val ROT_ABOUT_Y = register("rot_about_x", "aaqqqqqew", HexDir.SOUTH_WEST)
    val ROT_ABOUT_Z = register("rot_about_x", "aaqqqqqed", HexDir.SOUTH_WEST)
    val CONSTRUCT_ABOUT_X = register("cons_about_x", "daqqqqqea", HexDir.NORTH_WEST)
    val CONSTRUCT_ABOUT_Y = register("cons_about_y", "daqqqqqew", HexDir.NORTH_WEST)
    val CONSTRUCT_ABOUT_Z = register("cons_about_z", "daqqqqqed", HexDir.NORTH_WEST)
    val NORMALIZE = register("normalize", "eeeeedww", HexDir.SOUTH_WEST)
    val LEN_EQ = register("len_eq", "adqqaqw", HexDir.EAST)
    val LEN_NEQ = register("len_neq", "daeedew", HexDir.EAST)
    val IN_RANGE = register("in_range", "qqqq", HexDir.SOUTH_WEST)
    val OUT_RANGE = register("out_range", "eaae", HexDir.SOUTH_EAST)
    val INVERT = register("invert", "waqawqa", HexDir.SOUTH_WEST)
    val INCREMENT = register("increment", "waawawaaw", HexDir.NORTH_EAST)
    val DECREMENT = register("decrement", "wddwdwddw", HexDir.NORTH_WEST)
    val APPROACH = register("approach", "dedqadeeed", HexDir.SOUTH_WEST)
    val ANGLE_DIST = register("angle_dist", "awdaqqqqqea", HexDir.NORTH_EAST)
    val ANGLE_APPROACH = register("angle_approach", "awdaqqqqqwd", HexDir.NORTH_EAST)
    val SET_INSERT_RET = register("set_insert_ret", "edqdewd", HexDir.SOUTH_WEST )
    val SET_REMOVE_RET = register("set_remove_ret", "edqdewaqaaed", HexDir.SOUTH_WEST)
     */

    fun init(){
        register("true_eval", "deaqqaaqa", HexDir.SOUTH_EAST, OpConditionalEval(true))
        register("false_eval", "deaqqdded", HexDir.SOUTH_EAST, OpConditionalEval(false))
        register("true_halt", "aqdeedded", HexDir.SOUTH_WEST, OpConditionalHalt(true))
        register("false_halt", "aqdeeaaqa", HexDir.SOUTH_WEST, OpConditionalHalt(false))
    }

    private fun register(name: String, signature: String, startDir: HexDir, action: Action){
        Registry.register(HexActions.REGISTRY, HextraPats.modLoc(name),
            ActionRegistryEntry(HexPattern.fromAngles(signature, startDir), action)
        )
    }

    private fun register(name: String, signature: String, startDir: HexDir): HexPattern {
        val pattern = HexPattern.fromAngles(signature, startDir)
        Registry.register(HexActions.REGISTRY, HextraPats.modLoc(name),
            ActionRegistryEntry(pattern, OperationAction(pattern))
        )
        return pattern
    }

    private fun <T : SpecialHandler> registerSpecialHandler(name: String, handler: SpecialHandler.Factory<T>){
        Registry.register(IXplatAbstractions.INSTANCE.specialHandlerRegistry, HextraPats.modLoc(name), handler);
    }

    private fun registerNoConsumeOp(name: String, signature: String, startDir: HexDir, copied: HexPattern, argc: Int){
        val pattern = HexPattern.fromAngles(signature, startDir)
        Registry.register(HexActions.REGISTRY, HextraPats.modLoc(name),
            ActionRegistryEntry(pattern, NoConsOperationAction(pattern, argc))
        )
    }
}
